### ISSUE TITLE
Provide correct and accurate Trilinos version info during development to help address changes in backward compatibility

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#4532]
CC: @trilinos/framework, @fryeguy52 

## Description

The current setup of Trilinos is that it contains a static version-controlled `Version.cmake` file with static version information:

```
SET(Trilinos_VERSION 12.13)
SET(Trilinos_MAJOR_VERSION 12)
SET(Trilinos_MAJOR_MINOR_VERSION 121300)
SET(Trilinos_VERSION_STRING "12.13 (Dev)")
SET(Trilinos_ENABLE_DEVELOPMENT_MODE_DEFAULT ON) # Change to 'OFF' for a release
```

This info is used to configure the `Trilinos_version.h` file in the build directory that is installed with the contents:

```
#define TRILINOS_MAJOR_VERSION 12
#define TRILINOS_MAJOR_MINOR_VERSION 121300
#define TRILINOS_VERSION_STRING "12.13 (Dev)"
```

The even/odd minor version number (13 currently) does tell the user if one is looking at an (odd-numbere) development version or an (even-numbered) release version but it does not give enough info for codes that use almost continuous integration with Trilinos (e.g. ASC Sierra, ATDM APPs, etc.).  For example, Trilinos breaks backward compatibility from time to time which causes downstream customer codes to fail to build or pass all of the tests. In a staged integration processes for customer APPs like ATDM SPARC (see [TRIL-243](https://sems-atlassian-son.sandia.gov/jira/browse/TRIL-243) and [TRIL-256](https://sems-atlassian-son.sandia.gov/jira/browse/TRIL-256)), the SPARC 'master' code needs to be able to build and pass tests against both their older accepted version of Trilinos and a newer Trilinos version on the 'develop' (or 'master') branch.  To accomplish this, when Trilinos 'develop' breaks backward compatibility, APP customer codes need to add ifdefs and other conditional logic to know what version of Trilinos they are pointing to.  In particular, they need to know if they are pointing a version of Trilinos before or after a particular version that breaks backward compatibility.  For example, they need to be able to write ifdefs like:

```
# if <current-trilinos-version> >= <known-version-that-breaks-x>
  # This is a newer version of Trilinos and we need to adjust for that change
  ...
#else
  # This is an older version of Trilinos were our current code works just fine
  ...
#endif
```

A Trilinos installation needs to be able to advertise its exact development version both at configure-time and at compile-time to enable logic like above.

Whatever solution is provided, it must be able to determine in simple C/C++ processor logic if a version of Trilinos is greater or less than some known version where a break in backward compatibility or a new feature is added.  It must work from direct builds of Trilinos and installations of Trilinos as well as from builds and installations created from source tarballs of Trilinos.

NOTE: Comparing git SHA1s is not an option because one must have access to the git repo in order compare versions with < or > comparisons and an installation of Trilinos does not have a full git repo.

## Proposed Solution

One way to provide the desired version info is to add the macro define:

```
#define TRILINOS_VERSION_DATE YYYYMMDDhh
```

which gives a monotonically increasing 10-digit integer for the version of Trilinos in terms of the 4-digit year (`YYYY`), 2-digit month (`MM`), 2-digit day (`DD`) and 2-digit hour (`hh`) that is taken from the git commit date (in UTC).

This 10-digit integer would fit in a 32-bit integer with a max signed value of `2^32/2 - 1` = `2147483647`.  Therefore, this could represent the Trilinos version every hour till the end of the year 2147 (or 128 years from now) with the last hour being `2147 12 31 23` = `2147123123`.  (Surely 64-bit integers will be handled in all C++ compilers in the year 2148).

This macro could be defined in a **new** file `Trilinos_version_date.h` that would not be included by any Trilinos code or tests (so as to not trigger any rebuilding of any Trilinos libraries or exectuables).  This file would only get included by client source files that needs to know the exact fine-grained Trilinos version.

This date would be generated automatically from the git commit date of first-parent commits on the Trilinos 'develop' branch which gives a monotonically increasing date/time stamp.  For example, the commit:

```
commit 930294d48f4661013265448c1f254a667c66bf0f
Merge: 6354c01 87054d8
Author: trilinos-autotester <trilinos-autotester@trilinos.org>
Date:   Fri Mar 1 19:34:15 2019 -0700

    Merge Pull Request #4530 from rppawlo/Trilinos/nox-remove-deprecated-code
    
    Automatically Merged using Trilinos Pull Request AutoTester
    PR Title: NOX: remove deprecated code
    PR Author: rppawlo
```

has the commit date in UTC:

```
$ env TZ=GMT git log --format="%cd" --date=iso-local -1 930294d
2019-03-02 02:34:15 +0000
```

This date returned from this git command `2019-03-02 02:34:15` could be read in and set the date/time integer:

```
#define TRILINOS_VERSION_DATE 201903020234
```

A 32-bit integer in the C/C++ processor will be able to handle date/time integers of this format and correctly do integer comparisons with them.

To accomplish this, one could configure the file `Version.cmake` into the build directory from the template `Trilinos/cmake/Version.cmake.in` and then put that `Version.cmake` file in the install directory, put it into the source tarball, and also install it into the installation directory from the source tarball.  (This would be exactly how this is done for the `<Project>RepoVersion.txt` file, see [here](https://tribits.org/doc/TribitsBuildReference.html#generating-a-project-repo-version-file).)  Then `Trilinos_version_date.h` would be generated from `TRILINOS_VERSION_DATE` in the generated `Version.cmake` file.

**Comment by mhoemmen** [comment]
We're not far from using the Semantic Versioning Standard, and we're considering a change to Trilinos' policy anyway, so we might as well go all the way and adopt the Standard.

**Comment by bartlettroscoe** [comment]
@mhoemmen said:

> The only change to the current (de jure) policy would be not skipping odd minor release numbers.

Pretty much.   The way to designate a development version is to use a tag like `trilinos-X.Y.0-dev.Q`.  According to the SemVer standard, the versions sort as:

* `trilinos-X.Y.0-dev.Q` < `trilinos-X.Y.0` < `trilinos-X.Y.1` ...

so `trilinos-X.Y.0-dev.Q` is the 'Q'th development version prior to the next release `trilinos-X.Y.0`.

And through crafty usage of git tags and `git describe`, we can automatically generate a SemVer compatible info for the develop branch that is unique for every commit on the branch and also provide correct version info for releases.  I will outline that next.

**Comment by bartlettroscoe** [comment]
FYI: Figured out how to get the date/time integer directly with git:

```
$ env TZ=GMT git log --format="%cd" --date=iso-local -1 930294d
2019-03-02 02:34:15 +0000
```

That date/time string `2019-03-02 02:34:15 +0000` will be easy to parse in raw CMake to produce the integer `201903020234`.  So we don't need Python, just an `execute_process()` command setting the env var `TZ=GMT`.

I updated the "Proposed Solution" section for this.

**Comment by bartlettroscoe** [comment]
An idea of version handling with git consistent with SemVer is outlined [here](https://docs.google.com/document/d/1uVQYI2cmNx09fDkHDA136yqDTqayhxqfvjFiuUue7wo/edit#heading=h.xav3ed16muxt).  The careful use of `git describe --match="<tag-base-name>*"` can be used to create these SemVer versions automatically.  First, note the two SemVer items https://semver.org/#spec-item-9:

> A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version. Identifiers MUST comprise only ASCII alphanumerics and hyphen [0-9A-Za-z-]. Identifiers MUST NOT be empty. Numeric identifiers MUST NOT include leading zeroes. Pre-release versions have a lower precedence than the associated normal version. A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version. Examples: 1.0.0-alpha, 1.0.0-alpha.1, 1.0.0-0.3.7, 1.0.0-x.7.z.92.

and https://semver.org/#spec-item-11:

> Precedence refers to how versions are compared to each other when ordered. Precedence MUST be calculated by separating the version into major, minor, patch and pre-release identifiers in that order (Build metadata does not figure into precedence). Precedence is determined by the first difference when comparing each of these identifiers from left to right as follows: Major, minor, and patch versions are always compared numerically. Example: 1.0.0 < 2.0.0 < 2.1.0 < 2.1.1. When major, minor, and patch are equal, a pre-release version has lower precedence than a normal version. Example: 1.0.0-alpha < 1.0.0. Precedence for two pre-release versions with the same major, minor, and patch version MUST be determined by comparing each dot separated identifier from left to right until a difference is found as follows: identifiers consisting of only digits are compared numerically and identifiers with letters or hyphens are compared lexically in ASCII sort order. Numeric identifiers always have lower precedence than non-numeric identifiers. A larger set of pre-release fields has a higher precedence than a smaller set, if all of the preceding identifiers are equal. Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0.

It looks like the more standard tag names `<prefix>-X.Y.Z` got adopted for tagging version for xSDK releases:

```
git tag | grep xsdk
xsdk-0.2.0
xsdk-0.2.0-rc1
xsdk-0.3.0-rc1
xsdk-0.4.0-rc1
```

(but the SemVer standard would require something more like `xsdk-0.4.0-rc.1`.)

So that suggests that moving to `trilinos-X.Y.0` would make sense for Trilinos as well.

The SemVer standard says that `trilinos-X.Y.0-dev.Q` is **before** the release `trilinos-X.Y.0`.  Therefore, `trilinos-X.Y.0-dev.Q` is the Qth development version leading up to `trilinos-X.Y.0`.

To see how to take advantage of this for Trilinos, I will demonstrate with the current development version and current release 12.12.

First, find the intersection for when the Trilinos 12.12 branch was taken off of `develop`:

```
$ git log -1 `git merge-base github/develop github/trilinos-release-12-12-branch`
commit d3b096f4f1a984291442564a93f03592dad859b1
Author: Roscoe A. Bartlett <rabartl@sandia.gov>
Date:   Tue Jul 18 20:10:55 2017 -0600

    Add checkout of develop branch and other improvements (#482)
```

Now, tag the start of Trilinos 12.13 development for that next release:

```
$ git tag -a \
  -m "Start of development of Trilinos leading up to version 12.13.0" \
  trilinos-12.13.0-dev \
  d3b096f4f1a984291442564a93f03592dad859b1
```

With the current top Trilinos commit on 'develop':

```
$ git log -1 github/develop
commit 930294d48f4661013265448c1f254a667c66bf0f
Merge: 6354c01 87054d8
Author: trilinos-autotester <trilinos-autotester@trilinos.org>
Date:   Fri Mar 1 19:34:15 2019 -0700

    Merge Pull Request #4530 from rppawlo/Trilinos/nox-remove-deprecated-code
    
    Automatically Merged using Trilinos Pull Request AutoTester
    PR Title: NOX: remove deprecated code
    PR Author: rppawlo
```

the described version is:

```
$ git describe --match="trilinos-12.13*"
trilinos-12.13.0-dev-7554-g930294d
```

That is very useful version identifier to the human eye.  But with very little manipulation in raw CMake, that string can be used to generate a SemVer-compatible version string `trilinos-12.13.0-dev.7554` that will sort before `trilinos-12.13.0` and therefore is **before** the first release `trilinos-12.13.0`.  So `trilinos-12.13.0-dev.7554` is the 7554'th development version of Trilinos after the last release leading up to the `trilinos-12.13.0` release.

And we can play the same game with the release before that.  Find the start of the Trilinos 12.12 development activity by first finding the location of the creation of the Trilinos 12.10 branch:

```
$ git log-short -1 `git merge-base github/develop github/trilinos-release-12-10-branch`

bfe7d03 "MueLu: Xpetra: fix macro usage in cmake files"
Author: Jonathan Hu <jhu@sandia.gov>
Date:   Tue Nov 8 14:25:25 2016 -0800 (2 years, 4 months ago)
```

With that, we tag the start of the development activity leading up to `trilinos-12.12.0`:

```
$ git tag -a \
  -m "Start of development of Trilinos leading up to version 12.12.0" \
  trilinos-12.12.0-dev bfe7d03
```

And we can tag the last release on that branch:

```
$ git tag -a -m "Trilinos release 12.12.1" trilinos-12.12.1 trilinos-release-12-12-1
```

Looking at the tip of the Trilinos 12.12 branch:

```
$ git describe --match="trilinos-12.12*" github/trilinos-release-12-12-branch 
trilinos-12.12.1-10-gc2ce113
```

it looks like there are 10 commits after the 12.12.1 release got tagged.  Why was there 10 commits made on the `trilinos-release-12-12-branch` **after** version 12.12.1 got tagged but there was never a 12.12.2 patch release put out?

**Comment by bartlettroscoe** [comment]
CC: @rppawlo, @ccober6  

@jwillenbring, okay, the inevitable has occurred and Trilinos has occurred and Trilinos has purposefully broken backward compatibility.  Therefore,  can we pull the trigger on this and do just the following:

1. Change to configure `Version.cmake` from `Version.cmake.in` and put `Version.cmake.in` under version control.
2. Add `TRILINOS_VERSION_DATE` to `Version.cmake.in` and `Trilinos_version.h.in` so that the configured and installed `Trilinos_version.h` will have `TRILINOS_VERSION_DATE` that SPARC (and other customers) can use to determine the exact version of Trilinos they are building against?

The other version stuff about tags, etc, mentioned above can be put off for a later discussion.

Okay?

**Comment by bartlettroscoe** [comment]
CC: @jwillenbring, @fryeguy52, @rppawlo 

The one issue with defining `TRILINOS_VERSION_DATE` is that if someone does a raw tarball download off of github, then there is no git info available in order to fill in `TRILINOS_VERSION_DATE`.  In cases where the base Trilinos directory did not have a `.git/` directory, we would have to leave `TRILINOS_VERSION_DATE` empty in the generated `Version.cmake` file and an undefined macro in the generated `Trilinos_version.h` file.  But that is easy to deal with and customer code can have logic like:

```
#if defined(TRILINOS_VERSION_DATE) && TRILINOS_VERSION_DATE >= 201903201323
#  define SPARC_TEMPUS_CREATE_STEPPER_CHANGE_201903201323
#elif TRILINOS_MAJOR_MINOR_VERSION >= 121400
#  define SPARC_TEMPUS_CREATE_STEPPER_CHANGE_201903201323
#else
#  undef SPARC_TEMPUS_CREATE_STEPPER_CHANGE_201903201323
#endif
```

So above is the logic exactly what SPARC could use to determine if it needed to change how it calls `Tempus::createStepper()` to address #4693 where `201903201323` if for the UTC commit date `2019-03-20 13:23:49 +0000` for the commit 9d4360d gotten from:

```
$ cd Trilinos/
$ env TZ=GMT git log --format="%cd" --date=iso-local -1 9d4360d 
2019-03-20 13:23:49 +0000
```

Make sense?

**Comment by rppawlo** [comment]
If I understand this right, the version date would normally be generated at configure time with a git query to the last commit? I'd really like to avoid every trilinos commit having to change the date file.

**Comment by bartlettroscoe** [comment]
@rppawlo asked:

> If I understand this right, the version date would normally be generated at configure time with a git query to the last commit? 

Correct.  The file `Version.cmake` in the build directory would get generated in the build directory from the version controlled file `Version.cmake.in`.

> I'd really like to avoid every trilinos commit having to change the date file.

Is the worry that every commit would trigger a rebuild?  To that I would say that only source files that need to know the version should include `Trilinos_version.h`, thereby reducing the cost of any rebuilds.  But I can see there are a bunch of "version.cpp" files in Trilinos that include this file.

To address that what if we could put `TRILINOS_VERSION_DATE` into a new `Trilinos_version_date.h` file and then only client code that needs to know the version at the fine-grained a level would include `Trilinos_version_date.h`?  That way, new commits would not trigger the rebuild of any Trilinos code?  Only downstream customer code that needs to know this version would include this file.  Would that address your concerns?

**Comment by bartlettroscoe** [comment]
@rppawlo said:

> I'd really like to avoid every trilinos commit having to change the date file.

@rppawlo, after some more investigation, you were right be be concerned.  The file `Trilinos_version.h` gets included all over the place in Trilinos so anything that updates the contents of that file would cause massive rebuilds.

So we would need to put  `TRILINOS_VERSION_DATE` into a new called something like `Trilinos_version_date.h` or thinking ahead `Trilinos_semver.h` that could also contain a SemVer compatible version string like `trilinos-X.Y.Z-dev.Q` (e.g. `trilinos-12.14.0-dev.7750`) or `trilinos-X.Y.Z` (e.g. `trilinos-12.14.2`), etc.   Since this would only impact new files not including in any Trilinos source files this approach would have no impact on current Trilinos compilation.  It would only allow customers to access the fine-grained version info they need for cases like #4693.

**Comment by rppawlo** [comment]
@bartlettroscoe - that sounds good! My main concerns were increased developer time and complexity of the commit process:
1.  Remember back in the casl days when a code added a version file that automatically updated itself every time you typed make? It forced a relink of the code base even if nothing changed and added 5 to 10 minutes of build time without changing any code. With cuda build and link times already high, I'm just trying to avoid this kind of thing.
2. Don't want to burden trilinos devs with yet another process such as running a script before every commit to log something and having to remember to commit that file.
3. Even if we automated this by adding extra commits to every PR, that adds extra baggage to the git history and slows down triage/bisection.

**Comment by bartlettroscoe** [comment]
@rppawlo said:

> Remember back in the casl days when a code added a version file that automatically updated itself every time you typed make? It forced a relink of the code base even if nothing changed and added 5 to 10 minutes of build time without changing any code. With cuda build and link times already high, I'm just trying to avoid this kind of thing.

Yes you are correct. Thanks for reminding me of that.  We would have remembered that pretty quickly after we saw unnecessary rebuilds.

> Don't want to burden trilinos devs with yet another process such as running a script before every commit to log something and having to remember to commit that file.

It will all be 100% automated of course.  Trilinos developers and existing users will not notice anything different (unless they need the `Trilinos_version_date.h` file like SPARC does right now).

> Even if we automated this by adding extra commits to every PR, that adds extra baggage to the git history and slows down triage/bisection.

This will not result in any extra git commits at all (except for the initial commits to implement this).

**Comment by bartlettroscoe** [comment]
The files we would add to the repo (though TriBITS) are:

```
Trilinos/cmake/tribits/core/package_arch/
    VersionDate.cmake.in
    Tribits_version_date.h.in
```

and these would be used to generate in the build dir:

```
<build_dir>/
    VersionDate.cmake
    Trilinos_version_date.h
```

and what would be installed is:

```
<install_dir>/include/Trilinos_version_date.h
```

**Comment by jwillenbring** [comment]
@bartlettroscoe and I talked through the above strategy and are in agreement that this is a reasonable approach to move forward with.

**Comment by mhoemmen** [comment]
Developers should exercise discipline to avoid exposing this header file to code that doesn't need it.  Trilinos developers have been too free about putting stuff in header files.

**Comment by jmgate** [comment]
Trying to make sure I understand this.  Will this require Trilinos developers to remember to modify and then stage `Trilinos_version_date.h` before every commit?

**Comment by bartlettroscoe** [comment]
@jmgate asked:

> Trying to make sure I understand this. Will this require Trilinos developers to remember to modify and then stage Trilinos_version_date.h before every commit?

No, as outlined [above](https://github.com/trilinos/Trilinos/issues/4532#issuecomment-475645492) it gets generated automatically.  I will post a PR to the TriBITS repo with automated tests that will make this concrete and address this an other questions.

**Comment by bartlettroscoe** [comment]
FYI: PR #4750 has the completed implementation of this.

**Comment by bartlettroscoe** [comment]
With PR #4750, the basics of this issue are complete for the driving use case in #4693 once it is merged.  But there are a few other things that would be nice to do as well at some point noted below:

* Add the `<REPO_NAME_UC>_VERSION_DATE` vars to the `<Project>Config.cmake` file and test that they are set correctly.  (This would allow easier determination of the Trilinos version at configure time by client CMake projects like SPARC.)

* Add a test for TribitsExampleProject that creates a dummy .git/ directory and then uses mockprogram.py as git to generate the git-related stuff like `<Project>RepoVersion.txt`, `VersionDate.cmake` and `<RepoName>_version_date.h` without having to actually clone any git repos.  (This would allow almost full testing the version date feature without actually needing to clone anything and more testing could exist in test/core/ instead of having to rely on test/ctest_driver/.)

* Add generated `VersionDate.cmake` files to source tarball generated from `make package_source` and then test configure, build, and test from the untarred source tarball and check that the `<RepoName>_version_date.h` files get generated correctly.  (This would make version date information available to clients that get Trilinos as a generated tarball.  Unfortunately, this will not work for tarballs created by GitHub so it is not clear how useful this feature would be.)

**Comment by bartlettroscoe** [comment]
NOTE: The semver idea has been transferred to the new issue #4752.  I have taken it out of the above "Proposed Solution" field but I will leave the comments since I can't move this to #4752.

**Comment by bartlettroscoe** [comment]
FYI: Because I merged this to 'atdm-nightly' yesterday, this ran in the ATDM Trilinos builds today.  We see the `Trilinos_version_date.h` file getting generated and installed now as shown by:

```
$ cat /projects/atdm_devops/trilinos_installs/2019-03-29/cee-rhel6_clang-5.0.1_openmpi-1.10.2_serial_static_opt/include/Trilinos_version_date.h 
/* TRILINOS_VERSION_DATE
 *
 * This macro gives the version date of the Trilinos git repository at
 * the time it was configured.  It gives the date YYYY-MM-DD and the 2-digit
 * (24) hour hh of the commit as a 10-digit integer:
 *
 *   YYYYYMMDDhh
 *
 * As long as the branch for the git repo is not hard reset, the first-parent
 * history should give a monotonically increasing 10-digit integer.  This
 * 10-digit date/time integer YYYYMMDDhh will fit in a signed 32-bit integer
 * with a maximum value of 2^32 / 2 - 1 = 2147483647.  Therefore, the maximum
 * date that can be handled is the year 2147 with the max date/time of 2147 12
 * 31 23 = 2147123123.  Modern C/C++ compilers (and other implementations of
 * the C preprocesor) should be able to handle 32-bit integers.
 * 
 * This header file is meant to be included by downstream codes to determine
 * the version of Trilinos being used and allows
 * TRILINOS_VERSION_DATE to be used in C/C++ ifdefs like:
 *
 * #if defined(TRILINOS_VERSION_DATE) && TRILINOS_VERSION_DATE >= 2019032704
 *   // The version is newer than 2019-03-27 04:00:00 UTC
 *   ...
 * #else
 *   // The version is older than 2019-03-27 04:00:00 UTC
 *     ...
 *  #endif
 *
 * This allows downstream codes to know the fine-grained version of
 * Trilinos at configure and build time to adjust for the addition of
 * new features, deprecation of code, or breaks in backward compatibility
 * (which occur in specific commits with unique commit dates).
 *
 */

#define TRILINOS_VERSION_DATE 2019032903
```

I think this is ready to be closed but it would be nice if someone would review this.

**Comment by bartlettroscoe** [comment]
This has been done and SPARC has used this twice to fix problem.

Closing as complete.
